src/audio/midi/MidiAudio.cpp

```diff
@@ -5,63 +5,38 @@
 #include "MidiAudio.hpp"
 #include "../terminal/parser/stateMachine.hpp"
 
-namespace
-{
-    class MidiOut
-    {
-    public:
-        static constexpr auto NOTE_OFF = 0x80;
-        static constexpr auto NOTE_ON = 0x90;
-        static constexpr auto PROGRAM_CHANGE = 0xC0;
+#include <dsound.h>
 
-        // We're using a square wave as an approximation of the sound that the
-        // original VT525 terminals might have produced. This is probably not
-        // quite right, but it works reasonably well.
-        static constexpr auto SQUARE_WAVE_SYNTH = 80;
+#pragma comment(lib, "dxguid.lib")
+#pragma comment(lib, "dsound.lib")
 
-        MidiOut() noexcept
-        {
-            if constexpr (Feature_DECPSViaMidiPlayer::IsEnabled())
-            {
-                midiOutOpen(&handle, MIDI_MAPPER, NULL, NULL, CALLBACK_NULL);
-                OutputMessage(PROGRAM_CHANGE, SQUARE_WAVE_SYNTH);
-            }
-        }
-        ~MidiOut() noexcept
-        {
-            if constexpr (Feature_DECPSViaMidiPlayer::IsEnabled())
-            {
-                midiOutClose(handle);
-            }
-        }
-        void OutputMessage(const int b1, const int b2, const int b3 = 0, const int b4 = 0) noexcept
-        {
-            if constexpr (Feature_DECPSViaMidiPlayer::IsEnabled())
-            {
-                midiOutShortMsg(handle, MAKELONG(MAKEWORD(b1, b2), MAKEWORD(b3, b4)));
-            }
-        }
+using Microsoft::WRL::ComPtr;
+using namespace std::chrono_literals;
 
-        MidiOut(const MidiOut&) = delete;
-        MidiOut(MidiOut&&) = delete;
-        MidiOut& operator=(const MidiOut&) = delete;
-        MidiOut& operator=(MidiOut&&) = delete;
+// The WAVE_DATA below is an 8-bit PCM encoding of a triangle wave form.
+// We just play this on repeat at varying frequencies to produce our notes.
+constexpr auto WAVE_SIZE = 16u;
+constexpr auto WAVE_DATA = std::array<byte, WAVE_SIZE>{ 128, 159, 191, 223, 255, 223, 191, 159, 128, 96, 64, 32, 0, 32, 64, 96 };
 
-    private:
-        HMIDIOUT handle = nullptr;
-    };
+MidiAudio::MidiAudio(HWND windowHandle)
+{
+    if (SUCCEEDED(DirectSoundCreate8(nullptr, &_directSound, nullptr)))
+    {
+        if (SUCCEEDED(_directSound->SetCooperativeLevel(windowHandle, DSSCL_NORMAL)))

```

Comment:
Can `ServiceLocator::LocateConsoleWindow()->GetWindowHandle()` return null?
Should we add the sample code [from the remarks](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/mt708948(v=vs.85)#remarks) here?

Reply:
> Can `ServiceLocator::LocateConsoleWindow()->GetWindowHandle()` return null?

I don't think so. Technically it could be null briefly during startup, but the `ConsoleAllocateConsole` call blocks until `InitWindowsSubsystem` is done (that's where the window gets created). And if the window couldn't be created for some reason, then the `ConsoleAllocateConsole` call would fail, and there simply wouldn't be a console.

In conpty mode it's different. In that case `ServiceLocator::LocateConsoleWindow` itself would return null, but we also would never get this far, because the `DECPS` sequence would be passed through to the conpty client instead.